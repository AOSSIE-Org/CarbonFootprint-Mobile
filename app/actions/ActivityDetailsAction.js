/*
 * Action creators to save activity data in Redux store
 */

export const SET_ACTIVITY_DATE = 'SET_ACTIVITY_DATE';
export const SET_ACTIVITY_START_TIME = 'SET_ACTIVITY_START_TIME';
export const SET_ACTIVITY_DURATION = 'SET_ACTIVITY_DURATION';
export const SET_ACTIVITY_SRC = 'SET_ACTIVITY_SRC';
export const SET_ACTIVITY_DEST = 'SET_ACTIVITY_DEST';
export const SET_ACTIVITY_TYPE = 'SET_ACTIVITY_TYPE';
export const SET_ACTIVITY_DISTANCE = 'SET_ACTIVITY_DISTANCE';
export const SET_ACTIVITY_CO2 = 'SET_ACTIVITY_CO2';

/**
 * Action Creator to set Activity Date
 * @param val date
 * @return {Object} action for SET_ACTIVITY_DATE
 */
export function setDate(val) {
    return {
        type: SET_ACTIVITY_DATE,
        value: val
    };
}

/**
 * Action Creator to set Activity Start Time
 * @param  val start time
 * @return {Object} action for SET_ACTIVITY_START_TIME
 */
export function setStartTime(val) {
    return {
        type: SET_ACTIVITY_START_TIME,
        value: val
    };
}

/**
 * Action Creator to set Activity Duration
 * @param val activity duration
 * @return {Object} action for SET_ACTIVITY_DURATION
 */
export function setDuration(val) {
    return {
        type: SET_ACTIVITY_DURATION,
        value: val
    };
}

/**
 * Action Creator to set Activity Source
 * @param val activity source
 * @return {Object} action for SET_ACTIVITY_SRC
 */
export function setSrc(val) {
    return {
        type: SET_ACTIVITY_SRC,
        value: val
    };
}

/**
 * Action Creator to set Activity Destination
 * @param val destination
 * @return {Object} action for SET_ACTIVITY_DEST
 */
export function setDest(val) {
    return {
        type: SET_ACTIVITY_DEST,
        value: val
    };
}

/**
 * Action Creator to set Activity Type
 * @param val activity type
 * @return {Object} action for SET_ACTIVITY_TYPE
 */
export function setType(val) {
    return {
        type: SET_ACTIVITY_TYPE,
        value: val
    };
}

/**
 * Action Creator to set Activity Distance
 * @param val distance
 * @return {Object} action for SET_ACTIVITY_DISTANCE
 */
export function setDistance(val) {
    return {
        type: SET_ACTIVITY_DISTANCE,
        value: val
    };
}

/**
 * Action Creator to set Activity CO2 in kg
 * @param val co2 in kg
 * @return {Object} action for SET_ACTIVITY_CO2
 */
export function setCO2(val) {
    return {
        type: SET_ACTIVITY_CO2,
        value: val
    };
}
