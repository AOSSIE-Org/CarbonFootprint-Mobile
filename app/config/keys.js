export const firebaseConfig = {
    apiKey: "AIzaSyAKKLtd_lBLfiiZKhA8HbbRERoYHjvAKh0",
    authDomain: "cgi-task.firebaseapp.com",
    databaseURL: "https://cgi-task.firebaseio.com",
    projectId: "cgi-task",
    storageBucket: "cgi-task.appspot.com",
    messagingSenderId: "317961617630"
};

export const googleSignInConfig = {
    clientID: null,
    scopes: ['openid', 'profile', 'email'],
    shouldFetchBasicProfile: true
};

export const twitterSignInConfig = {
    twitter_key: null,
    twitter_secret: null
};

export const googleRoadsAPIKey = null;
export const geocodingAPIKey = "AIzaSyBBF78K8cCM8DEI4y98s2LqvCqQ_X84O9o";
