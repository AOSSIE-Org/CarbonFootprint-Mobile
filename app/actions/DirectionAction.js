import { mapboxKey } from '../config/keys';
import crashlytics from '@react-native-firebase/crashlytics';

export const REQUEST_DIRECTION = 'REQUEST_DIRECTION';
export const RECEIVE_DIRECTION = 'RECEIVE_DIRECTION';
export const SET_SOURCE = 'SET_SOURCE';
export const SET_DESTINATION = 'SET_DESTINATION';
export const SET_REGION = 'SET_REGION';
export const SET_DISTANCE = 'SET_DISTANCE';
export const SET_DURATION = 'SET_DURATION';
export const NO_DIRECTION = 'NO_DIRECTION';

/**
 * Action Creator to set Source
 * @param json
 * @param name
 * @return {Object} action for SET_SOURCE
 */
export function set_source(json, name) {
    return {
        type: SET_SOURCE,
        source: json,
        sourceName: name
    };
}

/**
 * Action Creator to set Destination
 * @param  json
 * @param  name
 * @return {Object} action for SET_DESTINATION
 */
export function set_destination(json, name) {
    return {
        type: SET_DESTINATION,
        destination: json,
        destinationName: name
    };
}

/**
 * Action Creator to set region
 * @param  json
 * @return {Object} action for SET_REGION
 */
export function set_region(json) {
    return {
        type: SET_REGION,
        region: json
    };
}

/**
 * Action Creator to set Distance
 * @param json
 * @return {Object} action for SET_DISTANCE
 */
function set_distance(value) {
    return {
        type: SET_DISTANCE,
        distance: value
    };
}

/**
 * Action Creator to set Duration
 * @param json
 * @return {Object} action for SET_DURATION
 */
function set_duration(value) {
    return {
        type: SET_DURATION,
        duration: value
    };
}

/**
 * This is essentially a hack.
 * Need to take Earth's Spherical nature into account
 * @param  source      starting point of travel
 * @param  destination ending point of travel
 * @return {Promise}
 */
export function getRegion(source, destination) {
    return new Promise(function(resolve, reject) {
        let data = {};
        if (destination == null) {
            data = {
                latitude: source.latitude,
                longitude: source.longitude,
                latitudeDelta: 0.03,
                longitudeDelta: 0.03
            };
        } else {
            let minX = Math.min(source.latitude, destination.latitude);
            let minY = Math.min(source.longitude, destination.longitude);
            let maxX = Math.max(source.latitude, destination.latitude);
            let maxY = Math.max(source.longitude, destination.longitude);
            //alert("(maxX - minX): " + (maxX - minX) + ", (maxY - minY): " + (maxY - minY));
            data = {
                latitude: (minX + maxX) / 2,
                longitude: (minY + maxY) / 2,
                latitudeDelta: Math.max(maxX - minX + 0.1, 0.03),
                longitudeDelta: Math.max(maxY - minY + 0.1, 0.03)
            };
        }
        resolve(data);
    });
}

/**
 * Action Creator to request Direction
 * @return {Object} action for REQUEST_DIRECTION
 */
function request_direction() {
    return {
        type: REQUEST_DIRECTION
    };
}

/**
 * Action Creator to recieve Direction
 * @param json
 * @return {Object} action for RECEIVE_DIRECTION
 */
function receive_direction(json) {
    return {
        type: RECEIVE_DIRECTION,
        coords: json
    };
}

/**
 * Action Creator if Route is not Found
 * @return {Object} action for NO_DIRECTION
 */
function no_direction() {
    return {
        type: NO_DIRECTION
    };
}

/**
 * helper class to formatLocation
 * @param  x location
 * @return String latitude and longitude
 */
function formatLocation(x) {
    return x.latitude.toString() + ',' + x.longitude.toString();
}

export function getRedMarkerDetails(location, placename) {
    return dispatch => {
        getRegion({ latitude: location.latitude, longitude: location.longitude }, null).then(
            result => {
                dispatch(set_region(result));
                dispatch(set_source(location, placename));
            }
        );
    };
}

export function getGreenMarkerDetails(location, placename) {
    return dispatch => {
        dispatch(set_destination(location, placename));
    };
}

/**
 * Getting Directions From Google Maps API By Passing The Source,Destination And Code
 * @param String source start point of travel
 * @param String destination end point of travel
 * @param number code 1 for transit, 2 for bicycling, 3 for walking
 */
export function getDirections(source, destination, code) {
    return (dispatch, state) => {
        dispatch(request_direction());
        let start = source.longitude.toString() + ',' + source.latitude.toString();
        let end = destination.longitude.toString() + ',' + destination.latitude.toString();
        let mode = '';
        if (code === 1) {
            mode = 'driving/';
        } else if (code === 2) {
            mode = 'cycling/';
        } else if (code === 3) {
            mode = 'walking/';
        } else {
            mode = 'driving/';
        }

        return fetch(
            'https://api.mapbox.com/directions/v5/mapbox/' +
                mode +
                start +
                ';' +
                end +
                '?geometries=geojson&access_token=' +
                mapboxKey
        )
            .then(response => response.json())
            .then(json => {
                // Handle case for routes not found
                if (!('routes' in json) || json.routes.length <= 0) {
                    dispatch(no_direction());
                    getRegion(source, destination).then(result => {
                        dispatch(set_region(result));
                    });
                } else {
                    dispatch(set_distance(json.routes[0].distance));
                    dispatch(set_duration(json.routes[0].duration));
                    getRegion(source, destination).then(result => {
                        dispatch(set_region(result));
                        dispatch(receive_direction(json.routes[0].geometry));
                    });
                }
            })
            .catch(error => {
                crashlytics().log('Error while getting direction' + error.message);
            });
    };
}
