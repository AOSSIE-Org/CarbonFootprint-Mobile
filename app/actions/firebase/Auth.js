import auth from '@react-native-firebase/auth';
import { setUser, getUser } from './User';
import { urlToBase64 } from '../../config/helper';

/**
 * firebase function to register user
 * @param {string} name username
 * @param {string} email useremail
 * @param {string} password user password
 * @return {promise}
 */
export function registerFirebase(name, email, password) {
    return new Promise((resolve, reject) => {
        auth()
            .createUserWithEmailAndPassword(email, password)
            .then(() => {
                let temp = {
                    name,
                    email,
                    picture: null,
                    provider: 'email'
                };
                setUser(auth().currentUser.email, temp)
                    .then(() => {
                        getUser(auth().currentUser.email)
                            .then(user => {
                                resolve(user);
                            })
                            .catch(error => reject(error));
                    })
                    .catch(error => reject(error));
            })
            .catch(error => reject(error));
    });
}
/**
 * firebase function to login user
 * @param {string} email
 * @param {password} password
 * @return {Promise}
 */
export function loginEmailFirebase(email, password) {
    return new Promise((resolve, reject) => {
        auth()
            .signInWithEmailAndPassword(email, password)
            .then(user => {
                auth().onAuthStateChanged(function(user) {
                    if (user) {
                        user = user.toJSON();
                        getUser(user.email)
                            .then(user => resolve(user))
                            .catch(error => reject(error));
                    } else {
                        reject();
                    }
                });
            })
            .catch(error => reject(error));
    });
}

/**
 *
 * @param {string} type service used by user to login
 * @param {string} token from SERVICE API
 * @param {string} secret from SERVICE API
 * @return {Promise}
 */
export function loginCustomFirebase(type, token, secret) {
    return new Promise((resolve, reject) => {
        let credential = null;
        let provider = null;
        switch (type) {
            case 'facebook':
                // Facebook doesn't need a secret, rest all do.
                credential = auth.FacebookAuthProvider.credential(token);
                provider = 'facebook.com';
                break;
            default:
                //default is google
                credential = auth.GoogleAuthProvider.credential(token, secret);
                provider = 'google.com';
                break;
        }
        auth()
            .signInWithCredential(credential)
            .then(user => {
                user = user.user.toJSON();
                getUser(user.email)
                    .then(user => {
                        urlToBase64(user).then(user => {
                            resolve(user);
                        });
                    })
                    .catch(error => {
                        let userData = {
                            name: user.displayName,
                            email: user.email || null,
                            picture: user.photoURL || null,
                            provider: provider
                        };
                        urlToBase64(userData).then(userData => {
                            setUser(userData.email, userData)
                                .then(() => {
                                    getUser(userData.email)
                                        .then(user => resolve(user))
                                        .catch(error => reject(error));
                                })
                                .catch(error => reject(error));
                        });
                    });
            })
            .catch(error => {
                reject(error);
            });
    });
}

/**
 * firebase function to recover password by sending mail to user registered email
 * @param {string} email
 * @return {Promise}
 */
export function forgotPasswordFirebase(email) {
    return new Promise(function(resolve, reject) {
        auth()
            .sendPasswordResetEmail(email)
            .then(() => resolve())
            .catch(error => reject(error));
    });
}
