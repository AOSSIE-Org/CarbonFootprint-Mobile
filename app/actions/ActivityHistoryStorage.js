/*
 *  Functions to interact with Realm Db (local device storage)
 */

import Realm from 'realm';
import { setFootprint } from '../actions/firebase/Footprint';
import store from '../config/store';

/**
 * Activity History Storage System Using Realm And Firebase
 */
export default class ActivityHistoryStorage {
    /**
     * creating database for storing activity history
     */
    static createDB() {
        schema = {
            name: 'ActivityHistory',
            properties: {
                actDate: { type: 'string' },
                startTime: { type: 'string' },
                duration: { type: 'int' },
                src: { type: 'string' },
                dest: { type: 'string' },
                actType: { type: 'string' },
                distance: { type: 'double' },
                co2Emitted: { type: 'double' },
                co2Saved: { type: 'double' }
            }
        };
        this.realm = new Realm({ schema: [schema] });
    }

    /**
     * getting data on passed date
     * @param date
     * @return data of activity history on date
     */
    static getData(date) {
        return this.realm.objects('ActivityHistory').filtered('actDate = $0', date);
    }

    /**
     * getting total data on passed date
     * @param date
     * @return getting complete activity history data
     */
    static getTotalData(date) {
        var obj = this.realm.objects('ActivityHistory').filtered('actDate = $0', date);
        var len = obj.length;
        var co2Emitted = 0,
            co2Saved = 0,
            co2WalkSaved = 0,
            co2RunSaved = 0,
            co2CycleSaved = 0,
            co2VehicleEmitted = 0;
        var dist = 0,
            distWalk = 0,
            distRun = 0,
            distCycle = 0,
            distVehicle = 0;
        var dur = 0,
            durWalk = 0,
            durRun = 0,
            durCycle = 0,
            durVehicle = 0;
        for (var i = 0; i < len; i++) {
            co2Emitted = co2Emitted + obj[i].co2Emitted;
            co2Saved = co2Saved + obj[i].co2Saved;
            dist = dist + obj[i].distance;
            dur = dur + obj[i].duration;
            if (obj[i].actType === 'WALKING' || obj[i].actType === 'ON_FOOT') {
                co2WalkSaved = co2WalkSaved + obj[i].co2Saved;
                distWalk = distWalk + obj[i].distance;
                durWalk = durWalk + obj[i].duration;
            }
            if (obj[i].actType === 'RUNNING') {
                co2RunSaved = co2RunSaved + obj[i].co2Saved;
                distRun = distRun + obj[i].distance;
                durRun = durRun + obj[i].duration;
            }
            if (obj[i].actType === 'ON_BICYCLE') {
                co2CycleSaved = co2CycleSaved + obj[i].co2Saved;
                distCycle = distCycle + obj[i].distance;
                durCycle = durCycle + obj[i].duration;
            }
            if (obj[i].actType === 'IN_VEHICLE') {
                co2VehicleEmitted = co2VehicleEmitted + obj[i].co2Emitted;
                distVehicle = distVehicle + obj[i].distance;
                durVehicle = durVehicle + obj[i].duration;
            }
        }
        return {
            co2Emitted: co2Emitted,
            co2Saved: co2Saved,
            co2WalkSaved: co2WalkSaved,
            co2RunSaved: co2RunSaved,
            co2CycleSaved: co2CycleSaved,
            co2VehicleEmitted: co2VehicleEmitted,
            dist: dist,
            distWalk: distWalk,
            distRun: distRun,
            distCycle: distCycle,
            distVehicle: distVehicle,
            dur: dur,
            durWalk: durWalk,
            durRun: durRun,
            durCycle: durCycle,
            durVehicle: durVehicle
        };
    }

    /**
     * insert activity data in firebase
     */
    static insertDataInFirebase() {
        var dataObj = this.getTotalData(new Date().toDateString());
        var data = {
            total: {
                time: dataObj.dur,
                footprint: dataObj.co2Emitted,
                distance: dataObj.dist,
                co2Saved: dataObj.co2Saved
            },
            car: {
                time: dataObj.durVehicle,
                footprint: dataObj.co2VehicleEmitted,
                distance: dataObj.distVehicle
            },
            transit: {
                time: 0,
                footprint: 0.0,
                distance: 0.0
            },
            cycling: {
                time: dataObj.durCycle,
                footprint: dataObj.co2CycleSaved,
                distance: dataObj.distCycle
            },
            walking: {
                time: dataObj.durWalk,
                footprint: dataObj.co2WalkSaved,
                distance: dataObj.distWalk
            },
            running: {
                time: dataObj.durRun,
                footprint: dataObj.co2RunSaved,
                distance: dataObj.distRun
            }
        };
        setFootprint(data, store.getState().auth.user.email);
    }

    /**
     * insert data in local storage and storing in firebase
     * @param dataObj
     */
    static insertData(dataObj) {
        this.realm.write(() => {
            const hist = this.realm.create('ActivityHistory', dataObj);
            this.insertDataInFirebase();
        });
    }

    /**
     * deleting user data from local storage mostly when user logs out
     * @param dataObj
     */
    static deleteData(dataObj) {
        this.realm.write(() => {
            let obj = this.realm.create('ActivityHistory', dataObj);
            this.realm.delete(obj);
        });
    }
}
