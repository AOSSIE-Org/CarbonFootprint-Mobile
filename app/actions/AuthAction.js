import firebase from '@react-native-firebase/app';
import '@react-native-firebase/auth';
import '@react-native-firebase/database';
import { Actions, ActionConst } from 'react-native-router-flux';
import { initFirebase } from './firebase/Init';
import { registerFirebase, loginEmailFirebase, forgotPasswordFirebase } from './firebase/Auth';
import { RESET_PASSWORD } from '../config/constants';
import { formatEmail } from '../config/helper';
import { checkValidityForSignIn, redirectSignIn } from './firebase/Helper';
import { loaderToggle } from './LoaderAction';
import crashlytics from '@react-native-firebase/crashlytics';
import Toast from 'react-native-simple-toast';

export const REQUEST_AUTH = 'REQUEST_AUTH';
export const RECEIVE_AUTH = 'RECEIVE_AUTH';
export const RECEIVE_ERROR = 'RECEIVE_ERROR';
export const REQUEST_FORGOT = 'REQUEST_FORGOT';
export const RECEIVE_FORGOT = 'RECEIVE_FORGOT';

/**
 * action creator for request auth
 * @return {Object} action for REQUEST_AUTH
 */
export function requestAuth() {
    return {
        type: REQUEST_AUTH
    };
}

/**
 * action creator for recieve auth
 * @param json request fetched
 * @return {Object} action for RECEIVE_AUTH
 */
export function receiveAuth(json) {
    return {
        type: RECEIVE_AUTH,
        user: json
    };
}

/**
 * action creator when recieved error during auth
 * @param json error request fetched
 * @return {Object} action for RECEIVE_ERROR
 */
export function receiveError(json) {
    return {
        type: RECEIVE_ERROR,
        error: json
    };
}

/**
 * action creator to forgot request
 * @return {Object} action for REQUEST_FORGOT
 */
export function requestForgot() {
    return {
        type: REQUEST_FORGOT
    };
}

/**
 * action creator to handle forgot password
 * @param message
 * @return {Object} action for RECEIVE_FORGOT
 */
export function receiveForgot(message) {
    return {
        type: RECEIVE_FORGOT,
        message
    };
}

/**
 * login user in to app by using firebase
 * @param email useremail
 * @param password userpassword
 * @return login the user
 */
export function login(email, password) {
    return (dispatch, getState) => {
        dispatch(loaderToggle());
        checkValidityForSignIn(email, 'email')
            .then(result => {
                dispatch(requestAuth());
                loginEmailFirebase(email, password)
                    .then(user => {
                        if (firebase.auth().currentUser.emailVerified) {
                            dispatch(requestAuth());
                            dispatch(loaderToggle());
                            dispatch(receiveAuth(user));
                            Actions.main({ type: ActionConst.REPLACE });
                        } else {
                            dispatch(requestAuth());
                            dispatch(loaderToggle());
                            Toast.show('Please verify your email');
                        }
                    })
                    .catch(error => {
                        dispatch(loaderToggle());
                        dispatch(receiveError(error));
                    });
            })
            .catch(error => {
                dispatch(loaderToggle());
                if (error.code === 'REDIRECT_SIGN_IN') {
                    // email id is already registered with another account of a different auth provider
                    dispatch(redirectSignIn(error.userInfo.provider));
                } else {
                    dispatch(receiveError(error));
                }
            });
    };
}

/**
 * register the user using firebase in database
 * @param name username
 * @param email useremail
 * @param password userpassword
 * @return register the user
 */
export function register(name, email, password) {
    return (dispatch, getState) => {
        dispatch(loaderToggle());
        checkValidityForSignIn(email, 'email')
            .then(result => {
                dispatch(requestAuth());
                registerFirebase(name, email, password)
                    .then(user => {
                        firebase.auth().currentUser.sendEmailVerification();
                        dispatch(loaderToggle());
                        Actions.home({ type: ActionConst.REPLACE });
                        Toast.show('Verification Email has been sent');
                    })
                    .catch(error => {
                        dispatch(loaderToggle());
                        dispatch(receiveError(error));
                    });
            })
            .catch(error => {
                dispatch(loaderToggle());
                if (error.code === 'REDIRECT_SIGN_IN') {
                    // email id is already registered with another account of a different auth provider
                    dispatch(redirectSignIn(error.userInfo.provider));
                } else {
                    dispatch(receiveError(error));
                }
            });
    };
}

/**
 * getting auth details
 * @return recieving user auth
 */
export function initApp() {
    return (dispatch, getState) => {
        dispatch(requestAuth());
        initFirebase()
            .then(user => {
                dispatch(receiveAuth(user));
            })
            .catch(() => {
                dispatch(receiveAuth(null));
            });
    };
}

/**
 * forgot password action using firebase function
 * @param email user registered Email id
 * @return password reset
 */
export function forgotPassword(email) {
    return dispatch => {
        dispatch(requestForgot());
        forgotPasswordFirebase(email)
            .then(() => dispatch(receiveForgot(RESET_PASSWORD)))
            .catch(error => dispatch(receiveForgot(error.message)));
    };
}

/**
 * update action using firebase function
 * @return updates the user firebase
 */

export function updateUserFirebase(user) {
    return dispatch => {
        return new Promise((resolve, reject) => {
            const email = firebase.auth().currentUser.email;
            firebase
                .database()
                .ref('users/' + formatEmail(email))
                .update(user);
            dispatch(receiveAuth(user));
            resolve();
        }).catch(err => {
            crashlytics().log('Error while updating user data' + err.message);
        });
    };
}

/**
 * logout action using firebase function
 * @return logout the user
 */
export function logout() {
    return dispatch => {
        firebase
            .auth()
            .signOut()
            .then(() => {
                Actions.intro({ type: ActionConst.RESET });
                // Reset the store
                dispatch({
                    type: 'USER_LOGOUT'
                });
            })
            .catch(error => {
                crashlytics().recordError(error);
            });
    };
}
