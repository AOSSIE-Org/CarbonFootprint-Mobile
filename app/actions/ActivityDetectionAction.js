/*
 * Detecting user's activity and sending activity data to store in Realm Db
 */

import { Platform } from 'react-native';

import ActivityRecognition from 'react-native-activity-recognition';
import ActivityHistoryStorage from '../actions/ActivityHistoryStorage';
import BackgroundTimer from 'react-native-background-timer';
import {
    setDate,
    setStartTime,
    setDuration,
    setSrc,
    setType,
    setDistance,
    setCO2
} from '../actions/ActivityDetailsAction';
import { formatAMPM, getPlaceName } from '../config/helper';
import crashlytics from '@react-native-firebase/crashlytics';

/**
 * Start a timer that runs continuous 1000 milliseconds using react-native-background-timer
 * @param dispatch Dispatches an action. This is the only way to trigger a state change.
 * @param getState Returns the current state tree of your application.
 */

export default class ActivityDetection {
    static intervalId;
    static unsubscribe;

    static startTimer(dispatch, getState) {
        if (Platform.OS === 'android') {
            ActivityDetection.intervalId = BackgroundTimer.setInterval(() => {
                dispatch(setDuration(getState().activity.duration + 1));
            }, 1000);
        } else {
            // Not tested for iOS
            BackgroundTimer.start();
            setInterval(() => {
                dispatch(setDuration(getState().activity.duration + 1));
            }, 1000);
        }
    }

    static stopTimer() {
        if (Platform.OS === 'android') {
            BackgroundTimer.clearTimeout(ActivityDetection.intervalId);
        } else {
            BackgroundTimer.stop();
        }
    }

    static async sendDataForStorage(state) {
        var act = state.activity;
        var source = 'Source';
        var destin = 'Destination';
        if (act.src.latitude === -1) {
            crashlytics().log('Error in fetching location (source)');
        } else {
            await getPlaceName(act.src)
                .then(place => (source = place))
                .catch(error => {
                    crashlytics().log('ActivityDetectionAction (sendDataForStorage 1)' + error);
                });
        }
        if (act.dest.latitude === -1) {
            crashlytics().log('Error in fetching location (destination)');
        } else {
            await getPlaceName(act.dest)
                .then(place => (destin = place))
                .catch(error => {
                    crashlytics().log('ActivityDetectionAction (sendDataForStorage 2)' + error);
                });
        }
        var data = {
            actDate: act.date,
            startTime: act.startTime,
            duration: act.duration,
            src: source,
            dest: destin,
            actType: act.type,
            distance: act.distance,
            co2Emitted: act.type === 'IN_VEHICLE' ? act.co2 : 0,
            co2Saved: act.type === 'IN_VEHICLE' ? 0 : act.co2
        };
        /*console.log("Activity data sent for local storage. Date: " + data.actDate + ", Start time: " + data.startTime + ", Duration: " + 
        data.duration + ", Source: " + data.src + ", Destination: " + data.dest + ", Type: " + data.actType + 
        ", Distance: " + data.distance + ", co2 emitted: " + data.co2Emitted + ", co2 saved: " + data.co2Saved);*/
        ActivityHistoryStorage.insertData(data);
    }
    /**
     * user Activity Detection
     */
    static startActivityDetection() {
        return function(dispatch, getState) {
            ActivityHistoryStorage.createDB();
            // Interval (in ms) for Activity detection updates
            const detectionIntervalMillis = 100;
            if (Platform.OS === 'android') ActivityRecognition.start(detectionIntervalMillis);
            // Subscribe to updates
            ActivityDetection.unsubscribe = ActivityRecognition.subscribe(detectedActivities => {
                const mostProbableActivity = detectedActivities.sorted[0];
                var act = getState().activity;
                // If detected activity is different from ongoing activity
                if (mostProbableActivity.type !== act.type) {
                    if (
                        (Platform.OS === 'android' && mostProbableActivity.confidence >= 75) ||
                        Platform.OS === 'ios'
                    ) {
                        if (
                            act.type !== 'STILL' &&
                            act.type !== 'TILTING' &&
                            act.type !== 'UNKNOWN'
                        ) {
                            ActivityDetection.stopTimer();
                            ActivityDetection.sendDataForStorage(getState());
                        }
                        var currDate = new Date();
                        dispatch(setDate(currDate.toDateString()));
                        dispatch(setStartTime(formatAMPM(currDate)));
                        dispatch(setDuration(0));
                        dispatch(setSrc(act.dest));
                        dispatch(setType(mostProbableActivity.type));
                        dispatch(setDistance(0));
                        dispatch(setCO2(0));
                        if (
                            mostProbableActivity.type !== 'STILL' &&
                            mostProbableActivity.type !== 'TILTING' &&
                            mostProbableActivity.type !== 'UNKNOWN'
                        ) {
                            ActivityDetection.startTimer(dispatch, getState);
                        }
                    }
                }
            });
        };
    }
    /**
     * Stop activity detection and remove the listener
     */
    static closeActivityDetection() {
        ActivityRecognition.stop();
        ActivityDetection.unsubscribe();
    }
}
