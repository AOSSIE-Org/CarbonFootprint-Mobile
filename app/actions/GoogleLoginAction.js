/*
 *	Google Native Login Action
 */

import { GoogleSignin, statusCodes } from 'react-native-google-signin';
import { loaderToggle } from './LoaderAction';
import { Actions, ActionConst } from 'react-native-router-flux';
import { googleSignInConfig } from '../config/keys';
import * as actions from './AuthAction';
import { loginCustomFirebase } from './firebase/Auth';
import { KEYS_NOT_SET } from '../config/constants';
import { checkValidityForSignIn, redirectSignIn } from './firebase/Helper';
import crashlytics from '@react-native-firebase/crashlytics';

/**
 * google signin functionality to app
 * @return handling google login feature
 */
export function googleSignIn() {
    if (googleSignInConfig.webClientId === null) {
        return alert(KEYS_NOT_SET);
    }
    return async dispatch => {
        GoogleSignin.configure(googleSignInConfig);
        try {
            dispatch(loaderToggle());
            await GoogleSignin.hasPlayServices();
            const userInfo = await GoogleSignin.signIn();
            const tokens = await GoogleSignin.getTokens();
            await checkValidityForSignIn(userInfo.user.email, 'google.com');
            loginCustomFirebase('google', tokens.idToken, tokens.accessToken)
                .then(user => {
                    dispatch(actions.receiveAuth(user));
                    dispatch(loaderToggle());
                    Actions.main({
                        type: ActionConst.REPLACE
                    });
                })
                .catch(error => {
                    dispatch(loaderToggle());
                });
        } catch (error) {
            dispatch(loaderToggle());
            if (error.code === 'REDIRECT_SIGN_IN') {
                // email id is already registered with another account of a different auth provider
                dispatch(redirectSignIn(error.userInfo.provider));
            } else if (error.code === statusCodes.SIGN_IN_CANCELLED) {
                // user cancelled the login flow
            } else if (error.code === statusCodes.IN_PROGRESS) {
                // operation (f.e. sign in) is in progress already
            } else if (error.code === statusCodes.PLAY_SERVICES_NOT_AVAILABLE) {
                // play services not available or outdated
            } else {
                // some other error happened
                crashlytics().recordError(error);
            }
        }
    };
}
