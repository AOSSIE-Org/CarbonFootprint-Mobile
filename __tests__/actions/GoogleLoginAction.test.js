import { googleSignIn } from '../../app/actions/GoogleLoginAction';
import GoogleSignIn from 'react-native-google-sign-in';
import { Actions } from 'react-native-router-flux';
import { googleSignInConfig } from '../../app/config/keys';
import * as Auth from '../../app/actions/firebase/Auth';
import thunk from 'redux-thunk';
import logger from 'redux-logger';
import configureMockStore from 'redux-mock-store';

// Mocking store
const middlewares = [thunk, logger];
const mockStore = configureMockStore(middlewares);

// Testing googleSignIn() function in ./app/actions/GoogleLoginAction

describe('testing of googleSignIn', () => {
    let credentials = {
        idToken: 'someRandomStringOfIdToken',
        accessToken: 'someRandomStringOfAccessToken'
    };

    // Mocking functions used in googleSignIn function

    GoogleSignIn.configure = jest.fn((googleSignInConfig) => new Promise((resolve) => {
        resolve();
    }));
    GoogleSignIn.signInPromise = jest.fn(() => new Promise((resolve) => {
        resolve(credentials);
    }));
    Auth.loginCustomFirebase = jest.fn(() => new Promise((resolve) => {
        resolve(credentials);
    }));
    Actions.main = jest.fn();

    it('checks the working of google sign in', async() => {
        const store = mockStore({});
        expect.assertions(8);
        await store.dispatch(googleSignIn()).then(() => {
            // The following snapshot is stored in __snapshots__ folder
            expect(store.getActions()).toMatchSnapshot();
        });
        expect(GoogleSignIn.configure).toHaveBeenCalledTimes(1);
        expect(GoogleSignIn.configure).toHaveBeenCalledWith(googleSignInConfig);
        expect(GoogleSignIn.signInPromise).toHaveBeenCalledTimes(1);
        await GoogleSignIn.signInPromise().then((res) => expect(res).toEqual(credentials));
        expect(Auth.loginCustomFirebase).toHaveBeenCalledTimes(1);
        expect(Auth.loginCustomFirebase).toBeCalledWith('google', credentials.idToken, credentials.accessToken);
        expect(Actions.main).toHaveBeenCalledTimes(1);
    });
});