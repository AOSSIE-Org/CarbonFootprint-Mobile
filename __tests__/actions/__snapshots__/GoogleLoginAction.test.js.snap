// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing of googleSignIn checks the working of google sign in 1`] = `
Array [
  Object {
    "type": "RECEIVE_AUTH",
    "user": Object {
      "accessToken": "someRandomStringOfAccessToken",
      "idToken": "someRandomStringOfIdToken",
    },
  },
]
`;
